File: carddrawer/src/test/java/com/meli/android/carddrawer/configuration/DarkFontConfigurationTest.java
package com.meli.android.carddrawer.configuration;

import android.graphics.LinearGradient;
import android.support.v4.content.ContextCompat;
import android.text.TextPaint;

import com.meli.android.carddrawer.BasicRobolectricTest;
import com.meli.android.carddrawer.R;

import com.meli.android.carddrawer.configuration.shadow.ShadowConfiguration;
import com.meli.android.carddrawer.configuration.shadow.ShadowFontConfiguration;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static org.junit.Assert.assertEquals;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyFloat;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class DarkFontConfigurationTest extends BasicRobolectricTest {
    private DarkFontConfiguration darkFontConfiguration;

    @Before
    public void doBefore() {
        final ShadowConfiguration shadowFontConfiguration = new ShadowFontConfiguration(getContext());
        darkFontConfiguration = new DarkFontConfiguration(getContext(), shadowFontConfiguration);
    }

    @Test
    public void getColor_returnsValidColor() {
        int validColor = ContextCompat.getColor(getContext(), R.color.card_drawer_dark_font_empty_color);

        assertEquals(validColor, darkFontConfiguration.getColor());
    }

    @Test
    public void setShadow_callSetShadow() {
        TextPaint textPaint = mock(TextPaint.class);

        darkFontConfiguration.setShadow(textPaint);

        verify(textPaint).setShadowLayer(anyFloat(), anyFloat(), anyFloat(), anyInt());
    }

    @Test
    public void setShadow_DoesNotCallSetShadow() {
        DarkFontConfiguration darkNotShadowFontConfiguration = new DarkFontConfiguration(getContext());
        TextPaint textPaint = mock(TextPaint.class);

        darkNotShadowFontConfiguration.setShadow(textPaint);

        verify(textPaint, never()).setShadowLayer(anyFloat(), anyFloat(), anyFloat(), anyInt());
    }

    @Test
    public void setGradient_callsSetShaderWithGradient() {
        TextPaint textPaint = mock(TextPaint.class);

        darkFontConfiguration.setGradient(textPaint, 10, 10);

        verify(textPaint).setShader(any(LinearGradient.class));
    }

}

File: carddrawer/src/test/java/com/meli/android/carddrawer/configuration/DarkFontConfigurationTest.kt
package com.meli.android.carddrawer.configuration

import android.text.TextPaint
import androidx.core.content.ContextCompat
import com.meli.android.carddrawer.R
import com.meli.android.carddrawer.configuration.base.ConfigurationTestBase
import com.meli.android.carddrawer.configuration.shadow.ShadowConfiguration
import com.meli.android.carddrawer.configuration.shadow.ShadowFontConfiguration
import io.mockk.mockk
import io.mockk.verify
import org.junit.Test

class DarkFontConfigurationTest : ConfigurationTestBase() {

    private lateinit var darkFontConfiguration: DarkFontConfiguration

    private fun initWithShadow() {
        val shadowFontConfiguration: ShadowConfiguration = ShadowFontConfiguration(contextMock)
        darkFontConfiguration = DarkFontConfiguration(contextMock, shadowFontConfiguration)
    }

    @Test
    fun `when getting color with shadow font per parameter then return card_drawer_dark_font_empty_color`() {
        initWithShadow()
        val validColor = ContextCompat.getColor(contextMock, R.color.card_drawer_dark_font_empty_color)
        assert(darkFontConfiguration.color == validColor)
    }

    @Test
    fun `when call function setShadow then call setShadowLayer`() {
        initWithShadow()
        val textPaint = mockk<TextPaint>(relaxed = true)
        darkFontConfiguration.setShadow(textPaint)
        verify {
            textPaint.setShadowLayer(
                any(),
                any(),
                any(),
                any()
            )
        }
    }

    @Test
    fun`when getting color without shadow per parameter then return card_drawer_dark_font_empty_color`() {
        darkFontConfiguration = DarkFontConfiguration(contextMock)
        val validColor =
            ContextCompat.getColor(contextMock, R.color.card_drawer_dark_font_empty_color)
        assert(darkFontConfiguration.color == validColor)
    }

    @Test
    fun`when call setShadow without shadow per parameter then call setShadowLayer`() {
        val textPaint = mockk<TextPaint>()
        darkFontConfiguration = DarkFontConfiguration(contextMock)
        darkFontConfiguration.setShadow(textPaint)
        verify(exactly = 0) {
            textPaint.setShadowLayer(
                any(),
                any(),
                any(),
                any()
            )
        }
   }
}